server: unify and improve logging

mostlygeek opened the pull request:
This commit switches tsidp to slog and collapses HTTP error logging to a single function.

- use WARN/DEBUG level logging depending on http status code
- move tsnet.Server logging to new flag: -debug-tsnet
- move request/response logging to new flag: -debug-all-requests
- unify API error handling logic to writeHTTPError
    
Updates #25

Log output: 

```
2025/09/23 16:26:06 WARN HTTP error status=401 method=GET path=/ error=access_denied description="not available over funnel"
2025/09/23 16:26:35 WARN HTTP error status=401 method=GET path=/authorize error=access_denied description="not allowed over funnel"
2025/09/23 16:26:44 WARN HTTP error status=401 method=GET path=/authorize error=access_denied description="not allowed over funnel"
```

- 401, 403 are logged at a WARN level
- 500 logged at an ERR level
- all other HTTP errors logged at DEBUG level

mostlygeek committed:
tsidp-server.go,Makefile,server: switch to slog

Switch to slog for structured logging with appropriate log levels for
activity within the app.

- log all token api errors at WARN level
- move tsnet.Server logging to new flag: -debug-tsnet
- move request/response logging to new flag: -debug-all-requests
- unify API error handling logic to writeHTTPError
- switch funnel error to http.StatusUnauthorized
- update docker image to use new logging flags
- update README.md with new flags and env vars

Co-authored-by: Andrew Lytvynov <awly@tailscale.com>
Signed-off-by: Benson Wong <benson@tailscale.com>

awly commented on server/appcap.go:
```diff
@@ -91,13 +91,13 @@ func (s *IDPServer) addGrantAccessContext(handler http.HandlerFunc) http.Handler
 		who, err = s.lc.WhoIs(r.Context(), remoteAddr)
 		if err != nil {
-			http.Error(w, fmt.Sprintf("Error getting WhoIs: %v", err), http.StatusInternalServerError)
+			writeHTTPError(w, r, http.StatusInternalServerError, "server_error", fmt.Sprintf("Error getting WhoIs: %v", err))
```
these error codes (e.g. `server_error`) look repetitive and somewhat consistent.
could you define some constants for them?

mostlygeek replied:
I added constants for the handful of errorCodes used.

awly commented on server/appcap.go:
```diff
@@ -91,13 +91,13 @@ func (s *IDPServer) addGrantAccessContext(handler http.HandlerFunc) http.Handler
 		who, err = s.lc.WhoIs(r.Context(), remoteAddr)
 		if err != nil {
-			http.Error(w, fmt.Sprintf("Error getting WhoIs: %v", err), http.StatusInternalServerError)
+			writeHTTPError(w, r, http.StatusInternalServerError, "server_error", fmt.Sprintf("Error getting WhoIs: %v", err))
```
we might not want to leak the exact error message to the caller here.
I recommend we separate the user-facing error from the internal logged one, like we do in https://github.com/tailscale/tailscale/blob/21c1f4cbff7a76d868faae3edf9a14ad2e5220d4/tsweb/tsweb.go#L886

mostlygeek replied:
fixed.

awly commented on server/authorize.go:
```diff
@@ -33,13 +32,13 @@ func (s *IDPServer) serveAuthorize(w http.ResponseWriter, r *http.Request) {
 	redirectURI := uq.Get("redirect_uri")
 	if redirectURI == "" {
-		http.Error(w, "tsidp: must specify redirect_uri", http.StatusBadRequest)
+		writeHTTPError(w, r, http.StatusBadRequest, "invalid_request", "tsidp: must specify redirect_uri")
```
why are some error messages prefixed with "tsidp: " and others not?
this prefix might not mean much for the average user

mostlygeek replied:
thanks. this has been cleaned up. Old messages migrated from previous code.

awly commented on server/authorize.go:
```diff
@@ -78,8 +77,8 @@ func (s *IDPServer) serveAuthorize(w http.ResponseWriter, r *http.Request) {
 	if err != nil {
-		log.Printf("Error getting WhoIs: %v", err)
-		http.Error(w, err.Error(), http.StatusInternalServerError)
+		slog.Error("Error getting WhoIs", slog.Any("error", err))
```
is this log here needed? or will `writeHTTPError` log the error anyway?

awly commented on server/authorize.go:
```diff
@@ -78,8 +77,8 @@ func (s *IDPServer) serveAuthorize(w http.ResponseWriter, r *http.Request) {
-		log.Printf("Error getting WhoIs: %v", err)
-		http.Error(w, err.Error(), http.StatusInternalServerError)
+		slog.Error("Error getting WhoIs", slog.Any("error", err))
+		writeHTTPError(w, r, http.StatusInternalServerError, "server_error", err.Error())
```
nit: don't leak the exact error message into the response body here

awly commented on server/authorize.go:
```diff
@@ -178,7 +176,8 @@ func redirectAuthError(w http.ResponseWriter, r *http.Request, redirectURI, erro
 	if err != nil {
 		// If redirect URI is invalid, return error directly
-		http.Error(w, "invalid redirect_uri", http.StatusBadRequest)
+		slog.Warn("Invalid redirect_uri", slog.String("error", err.Error()))
```
is this separate log needed?

awly commented on server/clients.go:
```diff
@@ -142,7 +142,7 @@ func (s *IDPServer) serveClients(w http.ResponseWriter, r *http.Request) {
 
 	if !ok {
-		http.Error(w, "tsidp: client not found", http.StatusNotFound)
+		writeHTTPError(w, r, http.StatusNotFound, "not_found", "tsidp: client not found")
```
nit: to match a similar error message from `authorize.go`
```suggestion
		writeHTTPError(w, r, http.StatusNotFound, "not_found", "tsidp: invalid client ID")
```

awly commented on server/clients.go:
```diff
@@ -196,7 +196,7 @@ func (s *IDPServer) serveNewClient(w http.ResponseWriter, r *http.Request) {
 
 	if err := s.storeFunnelClientsLocked(); err != nil {
-		http.Error(w, fmt.Sprintf("tsidp: failed to store client: %v", err), http.StatusInternalServerError)
+		writeHTTPError(w, r, http.StatusInternalServerError, "server_error", fmt.Sprintf("tsidp: failed to store client: %v", err))
```
nit: don't leak the underlying error message in response body

awly commented on server/clients.go:
```diff
@@ -262,7 +262,7 @@ func (s *IDPServer) serveDeleteClient(w http.ResponseWriter, r *http.Request, cl
 
 	if err := s.storeFunnelClientsLocked(); err != nil {
-		http.Error(w, fmt.Sprintf("tsidp: failed to store clients: %v", err), http.StatusInternalServerError)
+		writeHTTPError(w, r, http.StatusInternalServerError, "server_error", fmt.Sprintf("tsidp: failed to store clients: %v", err))
```
nit: don't leak the underlying error message in response body

awly commented on server/oauth-metadata.go:
```diff
@@ -211,7 +211,7 @@ func (s *IDPServer) serveJWKS(w http.ResponseWriter, r *http.Request) {
 	sk, err := s.oidcPrivateKey()
 	if err != nil {
-		writeJSONError(w, http.StatusInternalServerError, "server_error", "internal server error")
+		writeHTTPError(w, r, http.StatusInternalServerError, "server_error", "internal server error")
```
meta-comment: here and in quite a few other places we don't log the actual `error` value. I imagine that value would be quite useful for debugging. consider logging all `error` values internally (without returning them in response bodies)

awly commented on server/server.go:
```diff
@@ -284,18 +284,20 @@ func (s *IDPServer) oidcPrivateKey() (*signingKey, error) {
 				return &sk, nil
 			} else {
-				log.Printf("Error unmarshaling key: %v", err)
+				slog.Info("Error unmarshaling oidc key", slog.Any("error", err))
```
should this be Warn or even Error level?

awly commented on server/server.go:
```diff
@@ -433,3 +435,50 @@ func ServeOnLocalTailscaled(ctx context.Context, lc *local.Client, st *ipnstate.
+	case strings.Contains(acceptHeader, "text/plain"):
+		w.Header().Set("Content-Type", "text/plain; charset=UTF-8")
+		fmt.Fprintf(w, "Error %d: %s - %s", statusCode, errorCode, errorDescription)
+	default:
```
seems odd to send back a JSON if the client didn't specify `Accept`.
should plaintext be the default, and JSON only when `Accept: application/json`?

mostlygeek replied:
Picking JSON is a better default since most requests to tsidp endpoints will be an Oauth client of some sort instead of a user.

awly replied:
this is based on your assumption that tsidp is mostly used by MCP servers.
well-behaved OAuth clients can set the correct `Accept` header, no need to penalize the users who just jump through tsidp in their browser to access some website.

awly commented on server/server.go:
```diff
@@ -433,3 +435,50 @@ func ServeOnLocalTailscaled(ctx context.Context, lc *local.Client, st *ipnstate.
+		fmt.Fprintf(w, "Error %d: %s - %s", statusCode, errorCode, errorDescription)
+	default:
+		w.Header().Set("Content-Type", "application/json;charset=UTF-8")
+		json.NewEncoder(w).Encode(oauthErrorResponse{
```
why are all errors encoded as `oauthErrorResponse`? should we be doing it for all errors on all handlers?

mostlygeek replied:
legacy naming. I'll change it to something more abstract.

awly commented on server/server.go:
```diff
@@ -433,3 +435,50 @@ func ServeOnLocalTailscaled(ctx context.Context, lc *local.Client, st *ipnstate.
+type oauthErrorResponse struct {
+	Error            string `json:"error"`
+	ErrorDescription string `json:"error_description,omitempty"`
+	ErrorURI         string `json:"error_uri,omitempty"`
```
`ErrorURI` seems unused

awly commented on server/token.go:
```diff
@@ -155,12 +155,12 @@ func (s *IDPServer) serveToken(w http.ResponseWriter, r *http.Request) {
 		s.serveTokenExchange(w, r)
 	default:
-		writeTokenEndpointError(w, http.StatusBadRequest, "unsupported_grant_type", "")
+		writeHTTPError(w, r, http.StatusBadRequest, "unsupported_grant_type", "grant type sent: "+grantType)
```
don't echo a caller-provided value into the response, that can lead to XSS.

awly commented on server/token.go:
```diff
@@ -179,29 +179,28 @@ func (s *IDPServer) handleAuthorizationCodeGrant(w http.ResponseWriter, r *http.
 	s.mu.Unlock()
 	if !ok {
-		writeTokenEndpointError(w, http.StatusBadRequest, "invalid_grant", "code not found")
+		writeHTTPError(w, r, http.StatusBadRequest, "invalid_grant", "code not found")
```
should this be `StatusUnauthorized`?

awly commented on server/token.go:
```diff
@@ -259,8 +256,7 @@ func (s *IDPServer) handleRefreshTokenGrant(w http.ResponseWriter, r *http.Reque
 		if err != nil {
-			//log.Printf("Error validating resources: %v", err)
-			writeTokenEndpointError(w, http.StatusBadRequest, "invalid_request", err.Error())
+			writeHTTPError(w, r, http.StatusBadRequest, "invalid_request", err.Error())
```
should this be `StatusUnauthorized`?

mostlygeek replied:
`StatusBadRequest` seems to be fine since the client sent an invalid code.

awly commented on server/userinfo.go:
```diff
@@ -53,7 +53,7 @@ func (ui userInfo) toMap() map[string]any {
 func (s *IDPServer) serveUserInfo(w http.ResponseWriter, r *http.Request) {
 	if r.Method != "GET" {
-		http.Error(w, "tsidp: method not allowed", http.StatusMethodNotAllowed)
+		writeHTTPError(w, r, http.StatusMethodNotAllowed, "invalid_request", "tsidp: method not allowed")
```
optional: you can limit HTTP methods via `http.ServeMux` too.
if you change https://github.com/tailscale/tsidp/blob/5007403aa6fe3eb5f1851a94c419924c85fe7d55/server/server.go#L244
to something like
```go
	mux.HandleFunc("GET /userinfo", s.serveUserInfo)
```
it will automatically reject requests with the wrong method

mostlygeek replied:
Nice! I see this was a change in golang v1.22. There are tests that check for invalid methods. I'm OK keeping it the way it is. New code can make use of the new features in `mux`.

awly commented on server/userinfo.go:
```diff
@@ -97,7 +97,7 @@ func (s *IDPServer) serveUserInfo(w http.ResponseWriter, r *http.Request) {
 	rules, err := tailcfg.UnmarshalCapJSON[capRule](ar.RemoteUser.CapMap, tailcfg.PeerCapabilityTsIDP)
 	if err != nil {
-		http.Error(w, fmt.Sprintf("tsidp: failed to unmarshal capability: %v", err), http.StatusBadRequest)
+		writeHTTPError(w, r, http.StatusBadRequest, "invalid_request", fmt.Sprintf("tsidp: failed to unmarshal capability: %v", err))
```
this error message might leak the contents of `ar.RemoteUser.CapMap` to the caller. probably best to avoid it from the response body.

awly commented on tsidp-server.go:
```diff
@@ -54,7 +59,22 @@ func main() {
+		os.Exit(1)
+	}
+
+	switch *flagLogLevel {
+	case "debug":
+		slog.SetLogLoggerLevel(slog.LevelDebug)
+	case "info":
+		slog.SetLogLoggerLevel(slog.LevelInfo)
+	case "warn":
+		slog.SetLogLoggerLevel(slog.LevelWarn)
+	case "error":
+		slog.SetLogLoggerLevel(slog.LevelError)
+	default:
+		slog.Error("unknown log level", slog.String("level", *flagLogLevel))
+		os.Exit(1)
 	}
```
optional: you can also use https://pkg.go.dev/log/slog#Level.UnmarshalText

awly reviewed (commented):
Could you also add logging for successful authorizations through tsidp?
I'd like to have an audit log of everyone who used tsidp to successfully login to some remote service.

mostlygeek commented on server/clients.go:
```diff
@@ -423,22 +424,3 @@ func generateClientID() string {
 func generateClientSecret() string {
 	return rands.HexString(64)
 }
-
```
unused function - removed.

mostlygeek commented:
Hi @awly Thanks for the initial detailed review. I addressed the main issues: 

- don't leak error messages 
- replace errorCode strings with constants 
- cleaned up some nits and legacy messages
- removed some legacy code/conventions

awly commented on scripts/docker/run.sh:
```diff
@@ -27,16 +27,26 @@ if [ -n "$TSIDP_LOCAL_PORT" ]; then
+            ARGS="$ARGS -log=$TSIDP_LOG"
+            ;;
+        *)
+            echo "Error: TSIDP_LOG_LEVEL must be one of: debug, info, warn, error"
```
nit: I'd leave flag validation to the binary, no need to duplicate it here

mostlygeek replied:
I'd like to keep it here as the error is more specific to the docker container's env var setup.

awly commented on server/authorize.go:
```diff
@@ -134,13 +132,12 @@ func (s *IDPServer) serveAuthorize(w http.ResponseWriter, r *http.Request) {
 	parsedURL, err := url.Parse(redirectURI)
 	if err != nil {
-		http.Error(w, "invalid redirect URI", http.StatusInternalServerError)
+		writeHTTPError(w, r, http.StatusInternalServerError, ecServerError, "invalid redirect URI", nil)
```
```suggestion
		writeHTTPError(w, r, http.StatusInternalServerError, ecServerError, "invalid redirect URI", err)
```

awly commented on server/authorize.go:
```diff
@@ -178,7 +175,7 @@ func redirectAuthError(w http.ResponseWriter, r *http.Request, redirectURI, erro
 	if err != nil {
 		// If redirect URI is invalid, return error directly
-		http.Error(w, "invalid redirect_uri", http.StatusBadRequest)
+		writeHTTPError(w, r, http.StatusBadRequest, ecInvalidRequest, "invalid redirect_uri", nil)
```
```suggestion
		writeHTTPError(w, r, http.StatusBadRequest, ecInvalidRequest, "invalid redirect_uri", err)
```

awly commented on server/clients.go:
```diff
@@ -317,13 +318,13 @@ func (s *IDPServer) serveDynamicClientRegistration(w http.ResponseWriter, r *htt
 
 	if err := json.NewDecoder(r.Body).Decode(&registrationRequest); err != nil {
-		writeJSONError(w, http.StatusBadRequest, "invalid_request", "invalid request body")
+		writeHTTPError(w, r, http.StatusBadRequest, ecInvalidRequest, "invalid request body", nil)
```
```suggestion
		writeHTTPError(w, r, http.StatusBadRequest, ecInvalidRequest, "invalid request body", err)
```

awly commented on server/clients.go:
```diff
@@ -371,7 +372,7 @@ func (s *IDPServer) serveDynamicClientRegistration(w http.ResponseWriter, r *htt
 
 	if err := s.storeFunnelClientsLocked(); err != nil {
-		writeJSONError(w, http.StatusInternalServerError, "server_error", "failed to store client")
+		writeHTTPError(w, r, http.StatusInternalServerError, ecServerError, "failed to store client", nil)
```
```suggestion
		writeHTTPError(w, r, http.StatusInternalServerError, ecServerError, "failed to store client", err)
```

awly commented on server/oauth-metadata.go:
```diff
@@ -150,7 +150,7 @@ func (s *IDPServer) serveOAuthMetadata(w http.ResponseWriter, r *http.Request) {
 		return
 	}
 	if r.URL.Path != "/.well-known/oauth-authorization-server" {
```
nit: this check is not necessary, it's already enforced via https://github.com/tailscale/tsidp/blob/4f4ca34a88c5d6b7be2412be8ab55233dfeee87c/server/server.go#L233
same for all other handlers that check `r.URL.Path`, unless it's some dynamic path

awly commented on server/server.go:
```diff
@@ -139,15 +139,24 @@ type ActorClaim struct {
 }
 
+// for use with writeHTTPError() errorCode parameter
+const (
```
nit: make a `type errorCode string` for these to get a little validation help from the type system

mostlygeek replied:
leaving it as `string` is fine. The `writeHTTPError` takes a string for errorCode and I want to leave it in case we want to add a one off error code.

awly commented on server/server.go:
```diff
@@ -284,18 +293,20 @@ func (s *IDPServer) oidcPrivateKey() (*signingKey, error) {
 		if err != nil {
-			log.Fatalf("Error marshaling key: %v", err)
+			slog.Error("Error marshaling key", slog.Any("error", err))
+			os.Exit(1)
 		}
 		if err := os.WriteFile(keyPath, b, 0600); err != nil {
-			log.Fatalf("Error writing key: %v", err)
+			slog.Error("Error writing oid key", slog.Any("error", err))
+			os.Exit(1)
```
curious: why exit instead of returning the error?

mostlygeek replied:
Keeping behaviour same as previous `log.Fatalf`. Trying to keep changes to just logging in this PR.

mostlygeek replied:
The code surrounding this does a lazy initialization of key material. The original's let it crash approach makes it easy to troubleshoot but it won't crash until the middle of a request. If its worth it we can refactor out the lazy so crashing/initialization happens during app startup rather than in the middle of a request. 

We should do that in a different PR however.

awly commented on server/server.go:
```diff
@@ -433,3 +444,66 @@ func ServeOnLocalTailscaled(ctx context.Context, lc *local.Client, st *ipnstate.
+		slog.Int("status", statusCode),
+		slog.String("method", r.Method),
+		slog.String("path", r.URL.Path),
+		slog.String("error", errorCode),
```
just to avoid conflict with the `err` argument added below
```suggestion
		slog.String("errorCode", errorCode),
```

awly commented on server/server.go:
```diff
@@ -433,3 +444,66 @@ func ServeOnLocalTailscaled(ctx context.Context, lc *local.Client, st *ipnstate.
+//   - r: *http.Request to inspect Accept header and method/path for logging
+//   - statusCode: HTTP status code (e.g., 400, 401, 403, 500)
+//   - errorCode: Unique error code, use constants: ecAccessDenied, ecInvalidRequest, etc.
+//   - errorDescription: Human-readable error description
```
```suggestion
//   - errorDescription: Human-readable error description sent in the response body
```

awly commented on server/token.go:
```diff
@@ -155,12 +155,12 @@ func (s *IDPServer) serveToken(w http.ResponseWriter, r *http.Request) {
 		s.serveTokenExchange(w, r)
 	default:
-		writeTokenEndpointError(w, http.StatusBadRequest, "unsupported_grant_type", "")
+		writeHTTPError(w, r, http.StatusBadRequest, ecUnsupportedGrant, "grant type sent: "+grantType, nil)
```
```suggestion
		writeHTTPError(w, r, http.StatusBadRequest, ecUnsupportedGrant, "unsupported grant type", fmt.Errorf("unsupported grant type %q", grantType))
```

awly commented on server/token.go:
```diff
@@ -179,29 +179,28 @@ func (s *IDPServer) handleAuthorizationCodeGrant(w http.ResponseWriter, r *http.
 	}
 	if ar.RedirectURI != r.FormValue("redirect_uri") {
-		writeTokenEndpointError(w, http.StatusBadRequest, "invalid_grant", "redirect_uri mismatch")
+		writeHTTPError(w, r, http.StatusBadRequest, ecInvalidGrant, "redirect_uri mismatch", nil)
```
optional: consider logging the URI for debugging
```suggestion
		writeHTTPError(w, r, http.StatusBadRequest, ecInvalidGrant, "redirect_uri mismatch", fmt.Errorf("got redirect_uri %q, want %q", r.FormValue("redirect_uri"), ar.RedirectURI))
```

awly replied:
it this really `ecInvalidGrant`, or should it be `ecInvalidRequest`?

awly commented on server/token.go:
```diff
@@ -212,23 +211,22 @@ func (s *IDPServer) handleAuthorizationCodeGrant(w http.ResponseWriter, r *http.
 		if err != nil {
-			//log.Printf("Error validating resources: %v", err)
-			writeTokenEndpointError(w, http.StatusBadRequest, "invalid_request", "invalid resource")
+			writeHTTPError(w, r, http.StatusBadRequest, ecInvalidRequest, "invalid resource", nil)
```
```suggestion
			writeHTTPError(w, r, http.StatusBadRequest, ecInvalidRequest, "invalid resource", err)
```

awly commented on server/token.go:
```diff
@@ -292,53 +288,53 @@ func (s *IDPServer) handleRefreshTokenGrant(w http.ResponseWriter, r *http.Reque
 	// Parse form data
 	if err := r.ParseForm(); err != nil {
-		writeTokenEndpointError(w, http.StatusBadRequest, "invalid_request", "failed to parse form")
+		writeHTTPError(w, r, http.StatusBadRequest, ecInvalidRequest, "failed to parse form", nil)
```
```suggestion
		writeHTTPError(w, r, http.StatusBadRequest, ecInvalidRequest, "failed to parse form", err)
```

awly commented on server/ui.go:
```diff
@@ -262,18 +262,18 @@ func (s *IDPServer) handleEditClient(w http.ResponseWriter, r *http.Request) {
 
 		if err := r.ParseForm(); err != nil {
-			http.Error(w, "Failed to parse form", http.StatusBadRequest)
+			writeHTTPError(w, r, http.StatusBadRequest, ecInvalidRequest, "Failed to parse form", nil)
```
```suggestion
			writeHTTPError(w, r, http.StatusBadRequest, ecInvalidRequest, "Failed to parse form", err)
```

awly commented on README.md:
please generally avoid bundling multiple unrelated changes in one PR.
it makes it harder to review and slower for you to merge.

mostlygeek replied:
Agreed. That is typically my preference too. The README, at this early stage in the project, I think we can be a bit more flexible on.

awly commented on README.md:
```diff
@@ -60,26 +71,23 @@ If you'd like to build tsidp and / or run it directly you can do the following:
+> [!IMPORTANT]
+> Access to the admin UI and dynamic client registration endpoints are **denied** by default.
+
+To access the admin UI and dynamic client registration endpoints an [Application capability grant](https://tailscale.com/kb/1537/grants-app-capabilities) must be set in the the [Tailscale console](https://login.tailscale.com/admin).
```
```suggestion
To access the admin UI and dynamic client registration endpoints an [Application capability grant](https://tailscale.com/kb/1537/grants-app-capabilities) must be set in the the [Tailscale console](https://login.tailscale.com/admin/acls/).
```

awly commented on tsidp-server.go:
```diff
@@ -107,29 +131,39 @@ func main() {
-			ts.Logf = log.Printf
+		if *flagDebugTSNet {
+			ts.Logf = func(format string, args ...any) {
+				cur := slog.SetLogLoggerLevel(slog.LevelDebug) // force debug if this option is on
+				slog.Debug(fmt.Sprintf(format, args...))
+				slog.SetLogLoggerLevel(cur)
```
this will flip the global logger level while a line is printed.
instead, create a dedicated `slog.Logger` to use in tsnet.

mostlygeek replied:
This was part of the examples in the slog documentation. I can create a separate logger instead if we prefer to do it that way instead.

mostlygeek replied:
I looked into this. I would prefer to keep it this way. 

Rationale: 

- slog uses an internal, unexported handler to format logging. No way to keep consistency with the rest of the logging.
- it's ugly but less ugly than all the logs being in two different formats 
- `-debug-tsnet` is meant for troubleshooting and is not a hot code path.

awly replied:
oh, this actually sets the log level for `log.Print*` functions, not for `slog.Debug`. I don't think it matters here at all.

mostlygeek replied:
Maybe? 

This is a work around for the situation where `-log info` (default), and `-debug-tsnet` is used but doesn't print anything because slog filtered out debug level logs.

awly replied:
ah, https://pkg.go.dev/log/slog#SetLogLoggerLevel says
> Before [SetDefault](https://pkg.go.dev/log/slog#SetDefault) is called, slog top-level logging functions call the default [log.Logger](https://pkg.go.dev/log#Logger). In that mode, SetLogLoggerLevel sets the minimum level for those calls.

so because your code never calls `slog.SetDefault`, this implicitly sets the slog log level too.

mostlygeek committed:
server: add log for token issued

Signed-off-by: Benson Wong <benson@tailscale.com>

mostlygeek committed:
server: change default http error content type to text/plain

Signed-off-by: Benson Wong <benson@tailscale.com>

mostlygeek committed:
server: remove ambiguity in writeHTTPError slog attr names

Signed-off-by: Benson Wong <benson@tailscale.com>

mostlygeek committed:
server: remove unnecessary path checks and tests

Signed-off-by: Benson Wong <benson@tailscale.com>

linux4life798 commented on README.md:
```diff
@@ -18,38 +18,49 @@
 
-1. Use an existing auth key or create a new auth key in the [Tailscale dashboard](https://login.tailscale.com/admin/settings/keys). Ensure you select an existing [tag](https://tailscale.com/kb/1068/tags) or create a new one.
+> [!TIP]
+> Replace `YOUR_TAILSCALE_AUTHKEY` with your Tailscale authentication key in the following commands:
```
`YOUR_TAILSCALE_AUTHKEY` doesn't appear below. It looks like the `TS_AUTHKEY` was removed.

linux4life798 commented on README.md:
```diff
@@ -89,23 +97,75 @@ This is a permissive grant that is suitable only for testing purposes:
+
+> [!NOTE] > `TS_STATE_DIR` and `TS_HOSTNAME` are legacy names. These will be replaced by `TSIDP_STATE_DIR` and `TSIDP_HOSTNAME` in the future.
+
+| Environment Variable                     | CLI flag                   |
```
Shouldn't this table include `TS_AUTHKEY` for completeness?
